Documentation on how to configure the receiver's REST Transport (#11876)
CDCgov/prime-reportstream: prime-router/docs/onboarding-users/transport/rest.md

# REST Transport
## Introduction

> ReportStream supports four popular transport protocols:  Simple Object Access Protocol (SOAP), Resource State Transfer (REST), Secure File Transfer Protocol (SFTP), and Azure Blob. The goal of this document is to explain how to configure REST Transport for STLTs (a/k/a receivers). As well as provide guidance on how: 1) ReportStream REST Transport works, 2) to prepare a key for authentication, 3) to configure receiver's REST transport, and 4) test/check that a connection is successful. Please see ABC document on how to configure other transports.

Reviewer: ABC document?

Reviewer 2: That's a placeholder for future documentation on the other transports. Possibly remove that line?

Reviewer: There's already existing documentation that it refers to on other transport types...

Reviewer 2: Is there a link we can add?

Reviewer 3: I would suggest just link the  relevant doc to the stated transport protocol in the first sentence (e.g just have "Simple Object Access Protocol (SOAP) be a link to the .md file).

Reviewer: ```suggestion
ReportStream supports four popular transport protocols:  Resource State Transfer (REST), [Simple Object Access Protocol (SOAP)](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/soap.md), [Secure File Transfer Protocol (SFTP)](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/sftp.md), and [Azure Blob](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/blob.md). The goal of this document is to explain how to configure REST Transport for STLTs (a/k/a receivers). As well as provide guidance on how: 1) ReportStream REST Transport works, 2) to prepare a key for authentication, 3) to configure receiver's REST transport, and 4) test/check that a connection is successful. Please see ABC document on how to configure other transports.
```

Reviewer: Moved REST to the beginning, too, since it's this document and thus won't have a link.

Author: Thank you for all of the comments.

Author: Fixed.

> ## Assumptions ([refer to SFTP transport documentation for this section](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/sftp.md))

Reviewer: Maybe it's just me, but I feel like it's weird to refer to a document for a different transport type for assumptions.

Reviewer 3: agreed. maybe just copy relevant information straight from the SFTP document to here?

Author: Did copied the whole assumptions from SFTP.mp to here.

## 1.) How ReportStream REST stransport works
When STLT message is ready to be sent, ReportStream uploads the message to STLT's ready queue folder found in Azure Blobstorage (e.g., la-phd.elr, ca-doh.elr, etc.) and then triggers the send queue to start the SendFunction.
<ol type="a">
    <li>Using receiver.TransportType, the SendFunction calls to specific transport service accordingly. In our case, the SendFunction calls to the RESTTransport service.</li>
    <li>Reading credential from Azure Vault, RESTTransport uses its Type to determine the authentication type (i.e., UserPass or UserApiKey).  Next, it uses the obtained key to authenticates with STLT and sends the message payload.</li>
</ol>

> ## 2.) How-to prepare a key for authentication

Reviewer: ```suggestion
## 2.) How to prepare a key for authentication
```

For authentication, you need to do both step a and step b.
<ol type="a">
<li>Generate the "Credential in JSON format" for authentication</li>

Currently, RESTTransport uses the one of the three options: i) UserPass, ii) UserApiKey with JKS, or iii) UserApiKey with two-legged credential type to authenticate and obtain Bearer token from STLT. User can use **primeCLI** command with credential-create option to generate the "Credential in JSON format" as given below.

> - With STLT's credential username and password given to us by STLT, user need to run the following command to generate the UserPass credential type object:

Reviewer: ```suggestion
- With STLT's credential username and password given to us by STLT, user needs to run the following command to generate the UserPass credential type object:
```

Author: Fixed.

        Command:
            ./prime credential-create --type UserPass --user <username> --pass <password>
        
        Output: 
            Credential in JSON format: 
            ===
            {"@type":"UserPass","user":"<username>","pass":"<password>"}
            ===

> - With STLT's credential username and API Key given to us by STLT, user need to run the following command to generate the UserApiKey credential type object:

Reviewer: ```suggestion
- With STLT's credential username and API Key given to us by STLT, user needs to run the following command to generate the UserApiKey credential type object:
```

Author: Fixed.

	
          Command:
              ./prime credential-create --type UserApiKey --apikey-user <username> --apikey <API Key>
        
          Output: 
              Credential in JSON format: 
              ===
              {"@type":"UserApiKey","user":"<username>","key":"<API Key>"}
              ===

> - With STLT's credential jks-Username, jks-File-Password, and jks-key-file given to us by STLT, user need to run the following command to generate the UserJks credential type object:

Reviewer: ```suggestion
- With STLT's credential jks-Username, jks-File-Password, and jks-key-file given to us by STLT, user needs to run the following command to generate the UserJks credential type object:
```

Author: Fixed.

	
        Command:
            ./prime credential-create --type UserJks --jks-use <JKS Username> --jks-file-pass <JKS File Password> --jks-file <PATH to JKS File>
            
        Output: 
            Credential in JSON format: 
            ===
            {"@type":"UserJks","user":"<JKS Username>", "jks":"<JKS Key>"}
            ===

<li>Create Azure Key Vault Secret for STLT</li>

> With CDC SU password, login into the **Microsoft Azure** and navigate to the **pdhprod-clientconfig** and select **Secrets** from the menu on the left.  You should see the following screenshot.  **Note: Make sure your prod VPN is on**.

Reviewer: Is VPN still necessary/does it still work with this? I swear I've done it without, and I know there have been some changes made regarding the VPN since I started.

Reviewer 3: you shouldn't need VPN if your IP is whitelisted in pdhprod-clientconfig. Currently we only need VPN to access the staging DB (no idea why that's the line)

Reviewer 2: Can you please correct the wording as you see fit. I'm not sure how I would best update that.

Reviewer: ```suggestion
With CDC SU password, login into the **Microsoft Azure** and navigate to the **pdhprod-clientconfig** and select **Secrets** from the menu on the left.  You should see the following screenshot.
```

Author: Yes, for production, we still need to turn on VPN.

![](./images/rest-pdhprod-clientconfig-secrets.png)

Next refer to the above image, click on the  [+ Generate/Import] as the red arrow is pointing to.  You should see the following screenshot. Please follow the instruction given in the image.

![](./images/rest-create-a-secrete-key-screenshot.png)

</ol>

## 3.) How-to configure the receiver's REST transport

> The receiver's RESTtransport includes the following fields:

Reviewer: ```suggestion
The receiver's RESTTransport includes the following fields:
```

Author: Fixed.

              parameters: {}
              headers:
                UPHN-INFOMAP: "{\"properties\":\"labClia=10DRPTSTRM,target=NYS,content=L,format=HL7\"\
                  }"
              type: "REST"

    c) See UserApiKey+Tow-legged RESTTransport setting Example below:

        FLEXION--ETOR-SERVICE-RECEIVER uses UserApiKey + Tow-legged authentication type:
        ================================================================================
            transport: !<REST>
              reportUrl: "https://sample.net/v1/etor/orders"
              authTokenUrl: "https://sample.net/v1/auth/token"
              authType: "two-legged"
              tlsKeystore: null
              parameters:
                scope: "report-stream"
                client_assertion: "client_secret"
              headers:
                senderLabName: "CDC PRIME REPORTSTREAM"
                RecordId: "header.reportFile.reportId"
                sourceLabName: "CDC PRIME REPORTSTREAM"
              type: "REST"

## 4. Final Step is to test/check the receiver's REST transport is connected successfully
Now that you have completed/created REST Transport setting please do the following:

> - create PR - which includes the receiver's setting code

Reviewer: ```suggestion
- Create PR - which includes the receiver's setting code
```

Author: Fixed.

- PR gets approved
- Create the setting for staging environment

Login staging OKTA account

    ./prime login --env staging

Create a setting in staging environment by running the primeCLI command below:
(e.g., Assumming you are creating the setting for OK-PHD.ELR with setting file name ok-phd-setting.yml)

>     ./prime multiple-settings set --env staging --input ./settins/staging/ok-phd-setting.yml

Reviewer: ```suggestion
    ./prime multiple-settings set --env staging --input ./settings/staging/ok-phd-setting.yml
```

# REST Transport
## Introduction

> ReportStream supports four popular transport protocols:  Resource State Transfer (REST), [Simple Object Access Protocol (SOAP)](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/soap.md), [Secure File Transfer Protocol (SFTP)](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/sftp.md), and [Azure Blob](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/blob.md). The goal of this document is to explain how to configure REST Transport for STLTs (a/k/a receivers). As well as provide guidance on how: 1) ReportStream REST Transport works, 2) to prepare a key for authentication, 3) to configure receiver's REST transport, and 4) test/check that a connection is successful.

Reviewer 4: ```suggestion
ReportStream supports four popular transport protocols:  Resource State Transfer (REST), [Simple Object Access Protocol (SOAP)](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/soap.md), [Secure File Transfer Protocol (SFTP)](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/sftp.md), and [Azure Blob](https://github.com/CDCgov/prime-reportstream/blob/master/prime-router/docs/onboarding-users/transport/blob.md). The goal of this document is to explain how to configure REST Transport for STLTs (a/k/a receivers). As well as provide guidance on how: 
1) ReportStream REST Transport works
2) Prepare a key for authentication, 
3) Configure receiver's REST transport
4) Test/check that a connection is successful
```
Just separating the bullet points into separate lines.

#### Assumptions
* First and foremost: This assumes you have credentials with the CDC and have been given access to the Azure portal. If you do not have credentials, and access to the Azure portal, stop here. You cannot go any further.
* This also assumes you have access to KeyBase, and are part of the prime_dev_ops team there.
* Also, we assume you have KeyBase installed, and have in mounted into /Volumes/. If you don't have it mounted as a drive, you can just drag and drop the files into KeyBase. If you feel you should have access to the prime_dev_ops team in KeyBase, contact your dev lead, or the dev ops team lead and request access.
* Finally, this assumes you are working on a Unix-style system such as Linux or MacOS. You can probably do this on Windows as well provided you have a conformant shell with putty and openssl installed. The Git Bash shell would probably do, or the Linux Subsystem for Windows would as well, but no promises are made.
* Have an active Okta admin account.

> ## 1.) How ReportStream REST stransport works

Author: Fixed, thanks

> When STLT message is ready to be sent, ReportStream uploads the message to STLT's ready queue folder found in Azure Blobstorage (e.g., la-phd.elr, ca-doh.elr, etc.) and then triggers the send queue to start the SendFunction.

Reviewer 4: ```suggestion
When a STLT message is ready to be sent, ReportStream uploads the message to the STLT's ready queue folder found in Azure Blobstorage (e.g., la-phd.elr, ca-doh.elr, etc.) and then triggers the send queue to start the SendFunction.
```

Author: Fixed.

<ol type="a">

>     <li>Using receiver.TransportType, the SendFunction calls to specific transport service accordingly. In our case, the SendFunction calls to the RESTTransport service.</li>

Reviewer 4: ```suggestion
    <li>Using receiver.TransportType, the SendFunction calls the specific transport service accordingly. In our case, the SendFunction calls the RESTTransport service.</li>
```

>     <li>Reading credential from Azure Vault, RESTTransport uses its Type to determine the authentication type (i.e., UserPass or UserApiKey).  Next, it uses the obtained key to authenticates with STLT and sends the message payload.</li>

Reviewer 4: ```suggestion
    <li>Reads credential from Azure Vault, RESTTransport uses its Type to determine the authentication type (i.e., UserPass or UserApiKey).  Next, it uses the obtained key to authenticate with the STLT and sends the message payload.</li>
```

Author: Fixed

</ol>

## 2.) How to prepare a key for authentication
For authentication, you need to do both step a and step b.
<ol type="a">
<li>Generate the "Credential in JSON format" for authentication</li>

> Currently, RESTTransport uses the one of the three options: i) UserPass, ii) UserApiKey with JKS, or iii) UserApiKey with two-legged credential type to authenticate and obtain Bearer token from STLT. User can use **primeCLI** command with credential-create option to generate the "Credential in JSON format" as given below.

Reviewer 4: ```suggestion
Currently, RESTTransport uses the one of the three options: 
* i) UserPass, 
* ii) UserApiKey with JKS, or 
* iii) UserApiKey with two-legged credential type to authenticate and obtain Bearer token from the STLT. 

User can use **primeCLI** command with credential-create option to generate the "Credential in JSON format" as given below.
```
Separating bullet points into separate lines for readability.

Author: Fixed

	
          Command:
              ./prime credential-create --type UserApiKey --apikey-user <username> --apikey <API Key>
        
          Output: 
              Credential in JSON format: 
              ===
              {"@type":"UserApiKey","user":"<username>","key":"<API Key>"}
              ===

- With STLT's credential jks-Username, jks-File-Password, and jks-key-file given to us by STLT, user needs to run the following command to generate the UserJks credential type object:
	
        Command:
            ./prime credential-create --type UserJks --jks-use <JKS Username> --jks-file-pass <JKS File Password> --jks-file <PATH to JKS File>
            
        Output: 
            Credential in JSON format: 
            ===
            {"@type":"UserJks","user":"<JKS Username>", "jks":"<JKS Key>"}
            ===

<li>Create Azure Key Vault Secret for STLT</li>

With CDC SU password, login into the **Microsoft Azure** and navigate to the **pdhprod-clientconfig** and select **Secrets** from the menu on the left.  You should see the following screenshot.

![](./images/rest-pdhprod-clientconfig-secrets.png)

> Next refer to the above image, click on the  [+ Generate/Import] as the red arrow is pointing to.  You should see the following screenshot. Please follow the instruction given in the image.

Author: Yes, naming convention is OrganizationName--Receiver.name (e.g., NY-PHD--ELR where NY-PHD is OrgnizationName and ELR is Receiver.name).

![](./images/rest-create-a-secrete-key-screenshot.png)

</ol>

## 3.) How to configure the receiver's REST transport
The receiver's RESTTransport includes the following fields:

> 

Reviewer 4: How is the secret we create in Azure used or referenced when we set-up the receiver's transport? Is it retrieved by looking at the receiver's name?

Author: It get referenced by OrganizationName, and Receiver.name (e.g., NY-PHD--ELR).

    - headers:        The map of headers to be sent in the message
    - type:           "REST"

    a) See UserPass RESTTransport setting Example below:
        
        OK-PHD--ELR uses UserPass authentication type:
        ==============================================
            transport: !<REST>
              reportUrl: "https://sample.health.ok.gov/api/document/hl7"
              authTokenUrl: "https://sample.health.ok.gov/api/auth/token"
              authType: null
              tlsKeystore: null
              parameters: {}
              headers:
                senderLabName: "CDC PRIME REPORTSTREAM"
                RecordId: "header.reportFile.reportId"
                sourceLabName: "CDC PRIME REPORTSTREAM"
              type: "REST"

    b) See UserApiKey+JKS TLS/SSL RESTTransport setting Example below:

        NY-PHD--ELR uses UserApiKey + JKS TLS/SSL authentication type:
        ==============================================================
            transport: !<REST>
              reportUrl: "https://samplehealth.ny.gov/services/uphn/openapi/V1.0/ECLRSPOST"
              authTokenUrl: "https://samplehealth.ny.gov/token"
              authType: null

>               tlsKeystore: "NY-PHD--ELR--JKS"

Reviewer 4: Where is this TLS key store? Is this something we need to set up? Where do we get this value from?

Author: TLS is JKS that store in Azure vault where we store other credentials.  Yes to setup we run ./primeCLI as give in create credential step here.
![Screenshot 2023-10-26 at 9 22 22 AM](https://github.com/CDCgov/prime-reportstream/assets/20068283/36d96bd5-f5f8-4eb7-b545-1b164375ac27)
